[POST] Why we should rewrite enzymejs to rtl and how to do that (#51)
allegro/blog: _posts/2021-12-13-why-we-should-rewrite-enzymejs-to-rtl-and-how-to-do-that.md

---
layout: post
title: "Why should we rewrite enzyme.js to react-testing-library and how to do that?"
author: [magdalena.mazur]
tags: [tech, frontend, javascript, jest, enzyme, react-testing-library, rtl, react]
---
## Introduction

> Everyone repeats like a mantra that tests are an indispensable element of development work. Is there something about it? Well, I need to admit that as a developer, I rather often want to skip the test writing stage. I assume I'm not the only one. I am aware that it's a mistake, even with testers on board. Effective and efficient testing of your own code leads to catching bugs at the moment of creating a functionality, as well as when changing something already existing. It cannot be questioned. Sometimes tests also help to understand how some long-unused functionality or component works. And that's a small bonus too.

Reviewer: I would stick to one version: I am or I'm

Author: I changed everywhere to I'm

> Everyone repeats like a mantra that tests are an indispensable element of development work. Is there something about it? Well, I need to admit that as a developer, I rather often want to skip the test writing stage. I assume I'm not the only one. I am aware that it's a mistake, even with testers on board. Effective and efficient testing of your own code leads to catching bugs at the moment of creating a functionality, as well as when changing something already existing. It cannot be questioned. Sometimes tests also help to understand how some long-unused functionality or component works. And that's a small bonus too.

Reviewer 2: Is there something about it? -> Is there anything to it?

> Everyone repeats like a mantra that tests are an indispensable element of development work. Is there something about it? Well, I need to admit that as a developer, I rather often want to skip the test writing stage. I assume I'm not the only one. I am aware that it's a mistake, even with testers on board. Effective and efficient testing of your own code leads to catching bugs at the moment of creating a functionality, as well as when changing something already existing. It cannot be questioned. Sometimes tests also help to understand how some long-unused functionality or component works. And that's a small bonus too.

Reviewer 2: Zamień proszę interpunkcję na ładną typograficzną: apostrof `’`, em-dash `—` itp.

> Everyone repeats like a mantra that tests are an indispensable element of development work. Is there something about it? Well, I need to admit that as a developer, I rather often want to skip the test writing stage. I assume I'm not the only one. I am aware that it's a mistake, even with testers on board. Effective and efficient testing of your own code leads to catching bugs at the moment of creating a functionality, as well as when changing something already existing. It cannot be questioned. Sometimes tests also help to understand how some long-unused functionality or component works. And that's a small bonus too.

Reviewer 2: Dla wygodniejszego review, połam proszę wiersze na długość 120 znaków

**Can a large project cope without testing? Probably so. But the number of errors and their severity will probably be much higher.** That’s why in Allegro Ads we pay attention to writing tests.

> We have a fairly large number of frameworks that are used to write tests in JavaScript. We can list tools like: [MochaJS](https://mochajs.org/), [Jest](https://jestjs.io/), [Karma](https://karma-runner.github.io), [Jasmine](https://jasmine.github.io/), [Cypress](https://www.cypress.io/). Some of them focus more on business value, others on technical details. There are also a range of libraries which solve different problems. However, in this article, we will pay special attention and compare two testing javascript libraries like [Enzyme.js](https://enzymejs.github.io/enzyme/) and [React Testing Library](https://testing-library.com/docs/react-testing-library/intro/) (RTL).

Reviewer: I would emphasize that React Testing Library is often called RTL because later on you are referencing it by that name and it is easy to miss just "(RTL)"

![npm trends screen](/img/articles/2021-12-13-why-we-should-rewrite-enzymejs-to-rtl-and-how-to-do-that/01.png)
###### Source [npm trends](https://www.npmtrends.com/@testing-library/react-vs-enzyme)

## EnzymeJS
From the enzyme documentation we can read a short description: *Enzyme is a JavaScript Testing utility for React that makes it easier to test your React Components' output. You can also manipulate, traverse, and in some ways simulate runtime given the output.*

Enzyme was released in **2015 by AirBnB**. And let's face it, Enzyme has gained a lot of popularity. Last year, the library was moved from AirBnB space to external github space as a separate independent organization. This step was intended to encourage the community to further develop the library.

![commit diffrences](/img/articles/2021-12-13-why-we-should-rewrite-enzymejs-to-rtl-and-how-to-do-that/02.png)

![commit diffrences](/img/articles/2021-12-13-why-we-should-rewrite-enzymejs-to-rtl-and-how-to-do-that/03.png)
###### Source [Enzyme github](https://github.com/enzymejs/enzyme/commit/b06750bb5f6248ceb9c3fae903a71d0747a420d6)

> When it comes to raw data, the latest version of **Enzyme 3.11.0** has a bundle size at **463.2kb (127.5 kB)** and was released on **20 December 2019**. Currently has 258 [open issues](https://github.com/enzymejs/enzyme/issues) and 31 open PR’s. 	

Reviewer: bundle size of

![bundlephobia enzyme sreen](/img/articles/2021-12-13-why-we-should-rewrite-enzymejs-to-rtl-and-how-to-do-that/04.png)
###### Source [Bundle Phobia](https://bundlephobia.com/package/enzyme@3.11.0)

Interesting fact: 

> - There are also Enzyme Adapter Packages for React. For example, for version 16, it also needs *enzyme-adapter-react-16* to be installed. Unfortunately, it does not support all features completely.

Reviewer: I think we need some explanation what is the purpose of this adapters. Also it is one of the drawbacks of Enzyme (you have to wait for adapter after new React version) which should be emphazised.

## React Testing Library (RTL)

RTL says about itself: *The React Testing Library is a very light-weight solution for testing React components. It provides light utility functions on top of react-dom and react-dom/test-utils, in a way that encourages better testing practices.*.

Released in **April 2018 by Kent C. Dodds**. The latest stable version **12.1.2** of RTL has a bundle size at **215.8kb (47.3 kB)** and was released in **October 2021**. Currently, a pre-release of v13.0.0-alpha is being prepared. Repository has 26 [open issues](https://github.com/enzymejs/enzyme/issues) and 2 open PR’s.  

![bundlephobia rtl screen](/img/articles/2021-12-13-why-we-should-rewrite-enzymejs-to-rtl-and-how-to-do-that/05.png)

###### Source [Bundle Phobia](https://bundlephobia.com/package/@testing-library/react@12.1.2)

Interesting facts: 

> - When you create a project by Create React App, you will get support for React Testing Library right away. 

Reviewer: also React is officially recommending RTL as testing library https://reactjs.org/docs/testing.html#tools

- Jest documentation links to the RTL website in the [*Testing Web Framework*](https://jestjs.io/docs/testing-frameworks#react) section.   

## Basic differences

The main differences between the two libraries that I can notice:
1. **Different approaches to testing** 
   

>    What distinguishes rtl is that we find elements by their labels or texts, almost like a user would. The point is to replicate and simulate the user's behavior. Rtl try to render components to the DOM and doesn’t give us access directly to the internal component. While Enzyme uses shallow rendering (or sometimes deep rendering) and encourages us to test the components instance using their state and props. Enzyme also allows full rendering of DOM but it is not the basic assumption of the library's operation.

Reviewer: tries

2. **Bundle size** 
   

>    Maybe this is not as priority as the first point, but it is worth paying attention to bundle size also. RTL has a two times smaller bundle.

Reviewer: as significant?

3. **Age**  
   

>    Enzyme 6 years old, RTL 3 years old

Reviewer: Enzyme *is* 6 years old, RTL *is* 3 years old

![enzyme and rtl dowloands graph](/img/articles/2021-12-13-why-we-should-rewrite-enzymejs-to-rtl-and-how-to-do-that/06.png)
###### Source [npm trends](https://www.npmtrends.com/@testing-library/react-vs-enzyme)

In the case of RTL, it is clear that the number of downloads is still growing strongly. On the other hand, it can be suspected that the highest peak of enzyme downloads is already behind it.

###### Reasons to change testing library

Well, let's be honest, Enzyme.js is a powerful library which helps us test our React Components in an easier way for a long time. But maybe.. too long? The first official release was in December 2015. Do not get me wrong, I don't want to say that everything old is bad. But the frontend world is changing constantly, just as the approach to testing it. 
I know that the Enzyme.js library hasn’t been deprecated yet, but there are a lot of reasons to switch your test library right now. **If you haven't read it I highly recommend you to take a look at this [Piotr Staniów article](https://www.piotrstaniow.pl/goodbye-enzyme)** (it takes about 10 minutes to read). 

> The main reasons for me to think about slowly rewriting your tests and skip to another library from this article are: 

Reviewer: skip -> switch?

- one maintains a developer, 

> - not keeping up with React changes. 

Reviewer: there are other significant reasons to consider:
* concise API (few methods in RTL vs lots of them in Enzyme)
* one library, multiple frameworks - with the same API you can test angular, vue etc. using tesiting-library/angular

## Case study of rewriting

Below I will present my own example created for the purposes of this article. It should present the minimal effort to rewrite one library to another.

### Component: 

```js
import React, { useState } from 'react';

export const Test = ({ defaultCounter = 0, header = 'Default header' }) => {
    const [counter, setCounter] = useState(defaultCounter)
    const increment = () => setCounter(counter + 1)
    const decrement = () => setCounter(counter - 1)

    return (
        <div>
            <h1>{header}</h1>
            <p className='counter'>{counter}</p>
            <button onClick={increment}>Increment counter</button>
            <button onClick={decrement}>Decrement counter</button>
        </div>
    )
};

export default Test;

```
### Enzyme example: 

```js
import React from 'react';
import { mount } from 'enzyme';

import { Test } from '../Test';

describe('<Test />', () => {
  it('should render default header', () => {
    // when
    const wrapper = mount(<Test />);

    // then
    expect(wrapper.text()).toContain('Default header');
  });
  it('should render header from props', () => {
    // when

>     const wrapper = mount(<Test header="My simply calculator" />);

Reviewer: simple

  });
});

```
## Enzyme and RTL in Allegro Ads

My team is currently working on the development of [Allegro Ads](https://allegro.pl/ads), which is an advertising tool for our ecommerce platform sellers. It is a whole panel for viewing statistics, displaying advertising campaigns. We provide sponsored offers for our allegro site, facebook, google and recently even for on the partner websites. Our sellers pay us for the opportunity to advertise their products, and at the same time we strive to ensure that these ads accurately reach people potentially interested in buying, thereby generating profits for sellers.

The main technology stack is standard and still pretty cool. The core of the project is written in React, and significant parts have been rewritten to Typescript. From the beginning, for testing purposes, we use Jest. In my opinion this is a very good and stable tool, easy to work with. It has quite a low entry threshold, even if someone hasn’t had too much experience with tests before. The main advantages of Jest for me are minimal configuration and quite good documentation. But for testing react components we currently have two libraries and I'll explain why in a moment.

> The 2.0. version of platform Allegro Ads was released about 6 years ago, with an enzyme library on a board. A year ago the team added a react testing library to package.json. We have both packages so far. **Recently we officially decided to slowly rewrite the enzyme library for good.** We added the rule to our docc (Declaration of Code Convention) that the new components are tested just in rtl, the old ones will be rewritten when existing parts of the code will be touched.

Reviewer: onboard

## Migration effects so far

Currently we have got **592 test files** (counting after the file names with “spec” phrase). In these files we can also find **709 of unique “describe”** usage for wrapping a series of tests, and **1796 unique test cases**. Half of tests are written only in jest, but a significant amount of tests for better testing use extra libraries. 

React testing library is used in **131 files**, which is around **22% of all test files**. 
Enzyme tests are in **137 files** which is around **23%**.  

> So the enzyme is still used more often. However, keep in mind that we have been using Enzyme in this project from the beginning and rtl for a year. So I'm quite stunned by these numbers. I did not think that the use of rtl would increase so significantly in a year. Of course, there is still a long way to completely eliminate Enzyme, especially among other project priorities, but now I'm convinced that small steps will be enough to do that.

Reviewer: Enzyme vs enzyme, it should be consistent

## Summary

> Hopefully this article showed you that rewriting a library from one to the other doesn't have to be difficult. As programmers, we will encounter such situations frequently. I think that the effort put into replacing old libraries with newer ones and keeping your code free of technological debt will bring tangible results for developers, project and, finally, users.

Reviewer: we are not rewriting library :)

Author: I've change this line, maybe it was a bad word. Do you think that I should change also the title? I assume that "Why we should rewrite" sounds good and shows that article is about changing libraries. Not about rewriting one. ;)

---

Everyone repeats like a mantra that tests are an indispensable element of development work.  Is there anything to it? 
Well, I need to admit that as a developer, I rather often want to skip the test writing stage. I assume I’m not the 
only one. I’m aware that it’s a mistake, even with testers on board. Effective and efficient testing of your own code
 leads to catching bugs at the moment of creating a functionality, as well as when changing something already existing. 
 It cannot be questioned. Sometimes tests also help to understand how some long—unused functionality or component works. 
 And that’s a small bonus too. **Can a large project cope without testing? Probably so. But the number of errors and 
 their severity will probably be much higher.** That’s why in Allegro Ads we pay attention to writing tests.

## Frameworks
We have a fairly large number of frameworks that are used to write tests in JavaScript. We can list tools like: 
[MochaJS](https://mochajs.org/), [Jest](https://jestjs.io/), [Karma](https://karma-runner.github.io), 
[Jasmine](https://jasmine.github.io/), [Cypress](https://www.cypress.io/). Some of them focus more on business value,

>  others on technical details. There are also a range of libraries which solve different problems. However, in this 

Author: I removed this sentence. It's not important here.

>  others on technical details. There are also a range of libraries which solve different problems. However, in this 

Author: Usunęłam! Racja. :)

 article, we will pay special attention and compare two testing javascript libraries like 
 [Enzyme.js](https://enzymejs.github.io/enzyme/) and [React Testing Library](https://testing-library.com/docs/react-testing-library/intro/) (often called RTL).

![npm trends screen](/img/articles/2021-12-13-why-we-should-rewrite-enzymejs-to-rtl-and-how-to-do-that/01.png)
###### Source [npm trends](https://www.npmtrends.com/@testing-library/react-vs-enzyme)

## EnzymeJS

> From the Enzyme documentation we can read a short description: *Enzyme is a JavaScript Testing utility for React 

Reviewer 3: racja, niepotrzebny

that makes it easier to test your React Components' output. You can also manipulate, traverse, and in some ways 
simulate runtime given the output.*

Enzyme was released in **2015 by AirBnB**. And let’s face it, Enzyme has gained a lot of popularity. Last year, 
the library was moved from AirBnB space to external github space as a separate independent organization. This step was 
intended to encourage the community to further develop the library.

![commit diffrences](/img/articles/2021-12-13-why-we-should-rewrite-enzymejs-to-rtl-and-how-to-do-that/02.png)

![commit diffrences](/img/articles/2021-12-13-why-we-should-rewrite-enzymejs-to-rtl-and-how-to-do-that/03.png)
###### Source [Enzyme github](https://github.com/enzymejs/enzyme/commit/b06750bb5f6248ceb9c3fae903a71d0747a420d6)

> When it comes to raw data, the latest version of **Enzyme 3.11.0** has a bundle size of **463.2kb (127.5 kB)** and was 

Reviewer 3: maybe we should add minified/minified+gzipped to the sizes to be explicit?

released on **20 December 2019**. Currently has 258 [open issues](https://github.com/enzymejs/enzyme/issues) and 31 
open PR’s. 	

![bundlephobia Enzyme sreen](/img/articles/2021-12-13-why-we-should-rewrite-enzymejs-to-rtl-and-how-to-do-that/04.png)
###### Source [Bundle Phobia](https://bundlephobia.com/package/enzyme@3.11.0)

> Interesting fact: 

Reviewer 3: I would emphasise this paragraph more, the existance of adapters is a big deal. There's no official adapter for React 17 yet (and it's been a year!), blocking all of its users, who don't want to use community version, from upgrading.

- There are also Enzyme Adapter Packages for React. So you need to install adapter package and its dependencies related 
to the version of React that you are using. For example, for version 16, it also needs *enzyme—adapter—react—16* to be 
installed. Unfortunately, it does not support all features completely. So for every new version of React, we have to 

> wait for the release of the new adapter.

Author: Wyrzuciłam to zdanie z "So for every new version", bo tego chyba można się domyślić. A to, że 16 nie wspiera wszystkiego to ważne info.

provides light utility functions on top of react—dom and react—dom/test—utils, in a way that encourages better testing 
practices.*.

Released in **April 2018 by Kent C. Dodds**. The latest stable version **12.1.2** of RTL has a bundle size of 
**215.8kb (47.3 kB)** and was released in **October 2021**. Currently, a pre—release of v13.0.0—alpha is being 
prepared. Repository has 26 [open issues](https://github.com/enzymejs/enzyme/issues) and 2 open PR’s.  

![bundlephobia RTL screen](/img/articles/2021-12-13-why-we-should-rewrite-enzymejs-to-rtl-and-how-to-do-that/05.png)

###### Source [Bundle Phobia](https://bundlephobia.com/package/@testing-library/react@12.1.2)

Interesting facts: 
- When you create a project by Create React App, you will get support for React Testing Library right away. 
- Jest documentation links to the RTL website in the [*Testing Web Framework*](https://jestjs.io/docs/testing-frameworks#react)
 section as well as React documentation in the [*Recommended Tools*](https://reactjs.org/docs/testing.html#tools).
- Testing—library allows us to test with the same API not only React but also others frameworks like Vue 

> (testing—library/vue), Angluar (tesiting—library/angular), Svelte (testing—library/svelte) etc. This could be some 

Reviewer 3: there's also dom-testing-library which is used by all four of them and allows writing tests without any framework whatsoever!

> (testing—library/vue), Angluar (tesiting—library/angular), Svelte (testing—library/svelte) etc. This could be some 

Author: Dojrzysz wszystko! Dzięki :-)

advantage for people working on several projects in different technologies.

## Basic differences

The main differences between the two libraries that I can notice:
1. **Different approaches to testing** 
   
   What distinguishes RTL is that we find elements by their labels or texts, almost like a user would. The point is to 

>    replicate and simulate the user’s behavior. Rtl tries to render components to the DOM and doesn’t give us access 

Reviewer 3: I think we want to capitalize that 😄 
```suggestion
   replicate and simulate the user’s behavior. RTL tries to render components to the DOM and doesn’t give us access 
```

   directly to the internal component. While Enzyme uses shallow rendering (or sometimes deep rendering) and 
   encourages us to test the components instance using their state and props. Enzyme also allows full rendering of DOM 
   but it is not the basic assumption of the library’s operation.
2. **Bundle size** 
   

>    Maybe this is not as significant as the first point, but it is worth paying attention to bundle size also. RTL has 

Reviewer 3: I don't think the size matters that much, it doesn't impact the bundle size of an app.

>    Maybe this is not as significant as the first point, but it is worth paying attention to bundle size also. RTL has 

Author: Zgodnie z sugestią Kamila wyrzuciłam ten podpunkt całkowicie.

###### Source [npm trends](https://www.npmtrends.com/@testing-library/react-vs-enzyme)

In the case of RTL, it is clear that the number of downloads is still growing strongly. On the other hand, it can 
be suspected that the highest peak of Enzyme downloads is already behind it.

###### Reasons to change testing library

Well, let’s be honest, Enzyme.js is a powerful library which helps us test our React Components in an easier way for a 
long time. But maybe.. too long? The first official release was in December 2015. Do not get me wrong, I don’t want to 
say that everything old is bad. But the frontend world is changing constantly, just as the approach to testing it. 
I know that the Enzyme.js library hasn’t been deprecated yet, but there are a lot of reasons to switch your test library 
right now. **If you haven’t read it I highly recommend you to take a look at this 
[Piotr Staniów article](https://www.piotrstaniow.pl/goodbye-enzyme)** (it takes about 10 minutes to read). 

The main reasons for me to think about slowly rewriting your tests and switch to another library from this article are: 

> - one maintains a developer, 

Author: Zmieniłam na: "one developer maintaining the entire package"

- not keeping up with React changes. 

> I think these are big blockers in using this tool on.

Author: To pewnie mój wymysł :D Usunęłam on.

## Case study of rewriting

Below I will present my own example created for the purposes of this article. It should present the minimal effort to 
rewrite one library to another.

### Component: 

```js
import React, { useState } from 'react';

export const Test = ({ defaultCounter = 0, header = 'Default header' }) => {
    const [counter, setCounter] = useState(defaultCounter)
    const increment = () => setCounter(counter + 1)
    const decrement = () => setCounter(counter - 1)

    return (
        <div>
            <h1>{header}</h1>
            <p className='counter'>{counter}</p>
            <button onClick={increment}>Increment counter</button>
            <button onClick={decrement}>Decrement counter</button>
        </div>
    )
};

> export default Test;

Reviewer 3: I see the named export is used in tests below, so the default one is not necessary

```
### Enzyme example: 

```js
import React from 'react';
import { mount } from 'enzyme';

import { Test } from '../Test';

describe('<Test />', () => {
  it('should render default header', () => {
    // when
    const wrapper = mount(<Test />);

    // then
    expect(wrapper.text()).toContain('Default header');
  });

>   it('should render header from props', () => {

Reviewer 3: it would be nice for readability to separate those tests with a new line

    // when
    const wrapper = mount(<Test header="My simple calculator" />);

    // then
    expect(wrapper.text()).toContain('My simple calculator');
    expect(wrapper.text()).not.toContain('Default header');
  });
  it('should render default defaultCounter', () => {
    // when
    const wrapper = mount(<Test />);

    // then
    expect(wrapper.find('.counter').text()).toBe('0');
  });
  it('should render defaultCounter from props', () => {
    // when
    const wrapper = mount(<Test defaultCounter={100} />);

    // then
    expect(wrapper.find('.counter').text()).toBe('100');
  });
  it('should increment conunter on click', () => {
    // given
    const wrapper = mount(<Test defaultCounter={100} />);
    const button = wrapper.find('button').at(0);

    // when
    button.simulate('click');

    // then
    expect(wrapper.find('.counter').text()).toBe('101');
  });

>   it('should decrement conunter on click', () => {

Reviewer 4: ```suggestion
  it('should decrement counter on click', () => {
```

    // given
    const wrapper = mount(<Test defaultCounter={100} />);
    const button = wrapper.find('button').at(1);

    // when
    button.simulate('click');

    // then
    expect(wrapper.find('.counter').text()).toBe('99');
  });
});

```

> ### What need to be change:

Reviewer 4: ```suggestion
### What needs to be changed:
```

    // when
    render(<Test />);

    // then
    expect(screen.getByText('0')).toBeInTheDocument();
  });
  it('should render defaultCounter from props', () => {
    // when
    render(<Test defaultCounter={100} />);

    // then
    expect(screen.getByText('100')).toBeInTheDocument();
  });
  it('should increment conunter on click', () => {
    // given
    render(<Test defaultCounter={100} />);

    const button = screen.getByRole('button', { name: 'Increment counter' });

    // when
    userEvent.click(button);

    // then
    expect(screen.getByText('101')).toBeInTheDocument();
  });
  it('should decrement conunter on click', () => {
    // given
    render(<Test defaultCounter={100} />);

    const button = screen.getByRole('button', { name: 'Decrement counter' });

    // when
    userEvent.click(button);

    // then
    expect(screen.getByText('99')).toBeInTheDocument();
  });
});

```
## Enzyme and RTL in Allegro Ads

My team is currently working on the development of [Allegro Ads](https://allegro.pl/ads), which is an advertising tool 
for our ecommerce platform sellers. It is a whole panel for viewing statistics, displaying advertising campaigns. We 

> provide sponsored offers for our allegro site, facebook, google and recently even for on the partner websites. Our 

Reviewer 4: ```suggestion
provide sponsored offers for our Allegro site, Facebook, Google and recently even the partner websites. Our 
```

sellers pay us for the opportunity to advertise their products, and at the same time we strive to ensure that these 
ads accurately reach people potentially interested in buying, thereby generating profits for sellers.

The main technology stack is standard and still pretty cool. The core of the project is written in React, and 

> significant parts have been rewritten to Typescript. From the beginning, for testing purposes, we use Jest. In 

Reviewer 3: We don't use Jest from the beginning 😄  I rewrote it from Mocha/Chai/Sinon about a year ago

Author: Oh, good to know that!!!

my opinion this is a very good and stable tool, easy to work with. It has quite a low entry threshold, even if someone 
hasn’t had too much experience with tests before. The main advantages of Jest for me are minimal configuration and quite 
good documentation. But for testing react components we currently have two libraries and I’ll explain why in a moment.

> The 2.0. version of platform Allegro Ads was released about 6 years ago, with an Enzyme library on a board. A year ago 

Reviewer 3: ```suggestion
The 2.0 version of Allegro Ads platform was released about 6 years ago, with an Enzyme library on a board. A year ago 
```

the team added a react testing library to package.json. We have both packages so far. **Recently we officially decided 
to slowly rewrite the Enzyme library for good.** We added the rule to our docc (Declaration of Code Convention) that 

> the new components are tested just in RTL, the old ones will be rewritten when existing parts of the code will be onboard.

Reviewer 4: W jakim znaczeniu używamy tutaj "onboard"?

Znam tylko 
>to give new employees the knowledge and skills they need to become effective members of an organization

Author: To akurat było zaproponowane w komentarzu w review. Wcześniej było touch. Zmieniłam na "the old ones will be rewritten during the development of specific parts"

## Migration effects so far

Currently we have got **592 test files** (counting after the file names with “spec” phrase). In these files we can 
also find **709 of unique “describe”** usage for wrapping a series of tests, and **1796 unique test cases**. Half of 

> tests are written only in jest, but a significant amount of tests for better testing use extra libraries. 

Reviewer 4: ```suggestion
tests are written only in Jest, but a significant amount of tests for better testing use extra libraries. 
```
Szczególnie tutaj śmiesznie by to brzmiało 😄 [link](https://dictionary.cambridge.org/dictionary/english/jest?q=in+jest)
> If you say something **in jest**, you are not serious about it

React testing library is used in **131 files**, which is around **22% of all test files**. 
Enzyme tests are in **137 files** which is around **23%**.  

So the Enzyme is still used more often. However, keep in mind that we have been using Enzyme in this project from the 
beginning and RTL for a year. So I’m quite stunned by these numbers. I did not think that the use of RTL would increase
 so significantly in a year. Of course, there is still a long way to completely eliminate Enzyme, especially among 
 other project priorities, but now I’m convinced that small steps will be enough to do that.

## Summary

> Hopefully this article showed you that switch from one to the other library doesn’t have to be difficult. As 

Reviewer 4: ```suggestion
Hopefully this article showed you that the switch from one to the other library doesn’t have to be difficult. As 
```
```suggestion
Hopefully this article showed you that switching from one to the other library doesn’t have to be difficult. As 
```

programmers, we will encounter such situations frequently. I think that the effort put into replacing old 
libraries with newer ones and keeping your code free of technological debt will bring tangible results for developers, 
project and, finally, users.

> I hope that I was able to show you why we prescribe the Enzyme to RTL, how we do it and at what stage we are. Probably 

Reviewer 4: ```suggestion
I hope that I was able to show you why we rewrite the Enzyme tests to RTL, how we do it and at what stage we are at. Probably 
```
Prescribe - najczęściej w znaczeniu lekarz pacjentowi leki [klik](https://dictionary.cambridge.org/dictionary/english/prescribe)

*Source [npm trends](https://www.npmtrends.com/@testing-library/react-vs-enzyme)*

## EnzymeJS
From the Enzyme documentation we can read a short description:

> Enzyme is a JavaScript Testing utility for React that makes it easier to test your React Components' output.
> You can also manipulate, traverse, and in some ways simulate runtime given the output.

Enzyme was released in **2015 by AirBnB**. And let’s face it, Enzyme has gained a lot of popularity. Last year, the
 library was moved from AirBnB space to external github space as a separate independent organization. This step was
  intended to encourage the community to further develop the library.

![commit diffrences](/img/articles/2021-12-13-why-we-should-rewrite-enzymejs-to-rtl-and-how-to-do-that/02.png)

![commit diffrences](/img/articles/2021-12-13-why-we-should-rewrite-enzymejs-to-rtl-and-how-to-do-that/03.png)
*Source [Enzyme github](https://github.com/enzymejs/enzyme/commit/b06750bb5f6248ceb9c3fae903a71d0747a420d6)*

When it comes to raw data, the latest version of **Enzyme 3.11.0** has a bundle size of **463.2kB minified (127.5 kB

>  minified + gzipped)** and was released on **20 December 2019**. Currently is 258

Reviewer 4: ```suggestion
 minified + gzipped)** and was released on **20 December 2019**. Currently it has 258
```

RTL says about itself:

> The React Testing Library is a very light—weight solution for testing React components. It provides light utility
 functions on top of react—dom and react—dom/test—utils, in a way that encourages better testing practices.

Released in **April 2018 by Kent C. Dodds**. The latest stable version **12.1.2** of RTL has a bundle size of **215.8kB
 minified (47.3kB minified + gzipped)** . The RTL repository has 26
  [open issues](https://github.com/enzymejs/enzyme/issues) and 2 open PR’s.

![bundlephobia RTL screen](/img/articles/2021-12-13-why-we-should-rewrite-enzymejs-to-rtl-and-how-to-do-that/05.png)
*Source [Bundle Phobia](https://bundlephobia.com/package/@testing-library/react@12.1.2)*

Interesting facts:
- When you create a project by Create React App, you will get support for React Testing Library right away.
- Jest documentation links to the RTL website in the
[*Testing Web Framework*](https://jestjs.io/docs/testing-frameworks#react) section as well as React documentation in the
 [*Recommended Tools*](https://reactjs.org/docs/testing.html#tools).
- Testing—library allows us to test with the same API not only React but also others frameworks like Vue

>  (testing—library/vue), Angular (tesiting—library/angular), Svelte (testing—library/svelte) etc. This could be some

Reviewer 4: ```suggestion
 (testing—library/vue), Angular (testing—library/angular), Svelte (testing—library/svelte) etc. This could be some
```

>  (testing—library/vue), Angular (tesiting—library/angular), Svelte (testing—library/svelte) etc. This could be some

Reviewer 4: W przypadku nazw bibliotek zostawiłabym raczej dywizy zamiast myślników, bo to się dosyć dziwnie renderuje: 
![Zrzut ekranu 2022-01-20 o 13 09 20](https://user-images.githubusercontent.com/22297591/150336281-7d6965a7-2aa9-4ffd-afca-0f30ebd32832.png)
Albo opakowała nazwy paczek w backticki.

EDIT: w sumie kursywa z L76 też wygląda przyzwoicie 🤔 

Zostawiam do decyzji Tobie, ale do prezentacji nazw paczek używałabym konsekwentnie jednego formatu.

Author: Rzeczywiście źle to wyglądało. Zamieniłam na "mniejsze" myślniki i opakowałam w backticki.

2. **Age**

   Enzyme is 6 years old, RTL is 3 years old

## Comparison of downloads

Below we can see a comparison of the number of downloads of both libraries in the last 5 years. Between August and
 September 2020, the number of RTL downloads exceeded the number of Enzyme downloads.

![Enzyme and RTL dowloands graph](/img/articles/2021-12-13-why-we-should-rewrite-enzymejs-to-rtl-and-how-to-do-that/06.png)
*Source [npm trends](https://www.npmtrends.com/@testing-library/react-vs-enzyme)*

In the case of RTL, it is clear that the number of downloads is still growing strongly. On the other hand, it can be
 suspected that the highest peak of Enzyme downloads is already behind.

## Reasons to change the testing library

Well, let’s be honest, Enzyme.js is a powerful library which has helped us test our React Components in an easier way
 for a long time. But maybe... too long? The first official release was in December 2015. Do not get me wrong, I don’t
  want to say that everything old is bad. But the frontend world is changing constantly, just as the approach to
   testing it. I know that the Enzyme.js library hasn’t been deprecated yet, but there are a lot of reasons to switch

>     your test library right now. **If you haven’t read it I highly recommend you to take a look at this
>      [Piotr Staniów article](https://www.piotrstaniow.pl/goodbye-enzyme)** (it takes about 10 minutes to read).

Reviewer 4: ```suggestion
    your test library right now. **If you haven’t read it I highly recommend you to take a look at [this article](https://www.piotrstaniow.pl/goodbye-enzyme) by Piotr Staniów.** (it takes about 10 minutes to read).
```

> The main reasons for me to think about slowly rewriting your tests and switch to another library from this article are:

Reviewer 4: ```suggestion
The main reasons for you to think about slowly rewriting your tests and switch to another library from this article are:
```
skoro _twoje testy_ to także powód do zastanowienia _dla ciebie_, czytelniku

```suggestion
The main reasons for me to think about slowly rewriting my tests and switch to another library from this article are:
```
Ewentualnie w drugą stronę, powód do zastanowienia _dla mnie_, co zrobić z _moimi testami_.

Author: Takie oczywiste, a jednak nie zauważyłam tego. Dzięki!

> - one developer maintaining the entire package,

Reviewer 4: ```suggestion
- one developer maintaining the entire Enzyme package,
```
Akapit wyżej jest o Enzymie, ale by to jeszcze podkreśliła, że switch _z Enzyme_ na coś innego.

- not keeping up with React changes.

> I think these are big blockers in using this tool.

Reviewer 4: Proponuję zastąpić czymś "I think"
```suggestion
I believe these are big blockers in using this tool.
```
```suggestion
If you ask me, these are big blockers in using this tool.
```
```suggestion
From my standpoint these are big blockers in using this tool.
```
```suggestion
As far as I'm concerned these are big blockers in using this tool.
```

    // then
    expect(wrapper.text()).toContain('My simple calculator');
    expect(wrapper.text()).not.toContain('Default header');
  });

  it('should render default defaultCounter', () => {
    // when
    const wrapper = mount(<Test />);

    // then
    expect(wrapper.find('.counter').text()).toBe('0');
  });

  it('should render defaultCounter from props', () => {
    // when
    const wrapper = mount(<Test defaultCounter={100} />);

    // then
    expect(wrapper.find('.counter').text()).toBe('100');
  });

  it('should increment counter on click', () => {
    // given
    const wrapper = mount(<Test defaultCounter={100} />);
    const button = wrapper.find('button').at(0);

    // when
    button.simulate('click');

    // then
    expect(wrapper.find('.counter').text()).toBe('101');
  });

  it('should decrement counter on click', () => {
    // given
    const wrapper = mount(<Test defaultCounter={100} />);
    const button = wrapper.find('button').at(1);

    // when
    button.simulate('click');

    // then
    expect(wrapper.find('.counter').text()).toBe('99');
  });
});
```

### What needs to be changed:

> - remove **mount** and use **render** instead
> - remove **wrapper** variable and use **screen** to get elements
> - replace **simulate(’click’)** with **userEvent**

Reviewer 4: ```suggestion
- remove `mount` and use `render` instead
- remove `wrapper` variable and use `screen` to get elements
- replace `simulate('click')` with `userEvent`
```

Proponuję backtick do prezentacji słów kluczowych z kodu. Pogrubienie nie zwróciło aż tak mojej uwagi. 
<img width="579" alt="Zrzut ekranu 2022-01-20 o 13 33 57" src="https://user-images.githubusercontent.com/22297591/150339844-c0346821-506c-4006-86cc-c872723dfb91.png">

    expect(screen.queryByText('Default header')).not.toBeInTheDocument();
  });

  it('should render default defaultCounter', () => {
    // when
    render(<Test />);

    // then
    expect(screen.getByText('0')).toBeInTheDocument();
  });

  it('should render defaultCounter from props', () => {
    // when
    render(<Test defaultCounter={100} />);

    // then
    expect(screen.getByText('100')).toBeInTheDocument();
  });

  it('should increment counter on click', () => {
    // given
    render(<Test defaultCounter={100} />);

    const button = screen.getByRole('button', { name: 'Increment counter' });

    // when
    userEvent.click(button);

    // then
    expect(screen.getByText('101')).toBeInTheDocument();
  });

  it('should decrement counter on click', () => {
    // given
    render(<Test defaultCounter={100} />);

    const button = screen.getByRole('button', { name: 'Decrement counter' });

    // when
    userEvent.click(button);

    // then
    expect(screen.getByText('99')).toBeInTheDocument();
  });
});

```

## Enzyme and RTL in Allegro Ads

My team is currently working on the development of [Allegro Ads](https://allegro.pl/ads), which is an advertising tool

>  for our ecommerce platform sellers. It is a whole panel for viewing statistics, displaying advertising campaigns. We

Reviewer 4: ```suggestion
 for our e-commerce platform sellers. It is a whole panel for viewing statistics, displaying advertising campaigns. We
```

  provide sponsored offers for our Allegro site, Facebook, Google and recently even the partner websites. Our sellers
   pay us for the opportunity to advertise their products, and at the same time we strive to ensure that these ads
    accurately reach people potentially interested in buying, thereby generating profits for sellers.

The main technology stack is standard and still pretty cool. The core of the project is written in React, and
 significant parts have been rewritten to Typescript. For testing purposes, we use Jest. In my opinion this is a very
  good and stable tool, easy to work with. It has quite a low entry threshold, even if someone hasn’t had too much
   experience with tests before. The main advantages of Jest for me are minimal configuration and quite good
    documentation. But for testing react components we currently have two libraries and I’ll explain why in a moment.

> The 2.0 version of platform Allegro Ads was released about 6 years ago, with an Enzyme library on a board. A year ago

Reviewer 4: ```suggestion
The 2.0 version of platform Allegro Ads was released about 6 years ago, with Enzyme on board. A year ago
```
Chyba bym usunęła przedimek, może nawet uprościła do samego enzyme. 

Na pokładzie - on board.

---

Everyone repeats like a mantra that tests are an indispensable element of development work. Is there anything to it?
 Well, I need to admit that as a developer, I rather often want to skip the test writing stage. I assume I’m not the
  only one. I’m aware that it’s a mistake, even with testers on board. Effective and efficient testing of your own
   code can help with catching bugs in new functionalities, as well as in changes to already existing ones. It cannot
    be questioned. Sometimes tests also help to understand how some long-unused functionality or component works. And
     that’s a small bonus too. **Can a large project cope without testing? Probably so. But the number of errors and
      their severity will probably be much higher.** That’s why in Allegro Ads we pay attention to writing tests.

## Frameworks
We have a fairly large number of frameworks that are used to write tests in JavaScript. We can list tools like:
 [MochaJS](https://mochajs.org/), [Jest](https://jestjs.io/), [Karma](https://karma-runner.github.io),
  [Jasmine](https://jasmine.github.io/), [Cypress](https://www.cypress.io/). However, in this article, we will pay

>    special attention and compare two testing javascript libraries: [Enzyme.js](https://enzymejs.github.io/enzyme/)

Reviewer: two testing javascript -> two javascript testing ?

    replicate and simulate the user’s behavior. RTL tries to render components to the DOM and doesn’t give us access
     directly to the internal component. While Enzyme uses shallow rendering (or sometimes deep rendering) and
      encourages us to test the components instance using its state and props. Enzyme also allows full rendering of DOM
       but it is not the basic assumption of the library’s operation.
2. **Age**

   Enzyme is 6 years old, RTL is 3 years old

## Comparison of downloads

Below we can see a comparison of the number of downloads of both libraries in the last 5 years. Between August and
 September 2020, the number of RTL downloads exceeded the number of Enzyme downloads.

![Enzyme and RTL dowloands graph](/img/articles/2021-12-13-why-we-should-rewrite-enzymejs-to-rtl-and-how-to-do-that/06.png)
*Source [npm trends](https://www.npmtrends.com/@testing-library/react-vs-enzyme)*

In the case of RTL, it is clear that the number of downloads is still growing strongly. On the other hand, it can be
 suspected that the highest peak of Enzyme downloads is already behind.

## Reasons to change the testing library

> Well, let’s be honest, Enzyme.js is a powerful library which has helped us test our React Components in an easier way

Reviewer 3: ```suggestion
Well, let’s be honest, Enzyme.js is a powerful library which has helped us test our React Components in an easy way
```

    // when
    userEvent.click(button);

    // then
    expect(screen.getByText('101')).toBeInTheDocument();
  });

  it('should decrement counter on click', () => {
    // given
    render(<Test defaultCounter={100} />);

    const button = screen.getByRole('button', { name: 'Decrement counter' });

    // when
    userEvent.click(button);

    // then
    expect(screen.getByText('99')).toBeInTheDocument();
  });
});

```

## Enzyme and RTL in Allegro Ads

My team is currently working on the development of [Allegro Ads](https://allegro.pl/ads), which is an advertising tool
 for our e-commerce platform sellers. It is a whole panel for viewing statistics, displaying advertising campaigns. We
  provide sponsored offers for our Allegro site, Facebook, Google and recently even the partner websites. Our sellers
   pay us for the opportunity to advertise their products, and at the same time we strive to ensure that these ads
    accurately reach people potentially interested in buying, thereby generating profits for sellers.

The main technology stack is standard and still pretty cool. The core of the project is written in React, and

>  significant parts have been rewritten to Typescript. For testing purposes, we use Jest. In my opinion this is a very

Reviewer 3: ```suggestion
 significant parts have been rewritten to TypeScript. For testing purposes, we use Jest. In my opinion this is a very
```

>   good and stable tool, easy to work with. It has quite a low entry threshold, even if someone hasn’t had too much

Reviewer 3: ```suggestion
  good and stable tool, easy to work with. It has quite a low entry barrier, even if someone hasn’t had too much
```
obie formy są poprawne ale ja częściej spotykałem się z "barrier"

   experience with tests before. The main advantages of Jest for me are minimal configuration and quite good

>     documentation. But for testing react components we currently have two libraries and I’ll explain why in a moment.

Reviewer: React is everywhere else spelled with capital `R`

The 2.0 version of platform Allegro Ads was released about 6 years ago. A year ago

>  the team added a React Testing Library to `package.json`. We have both packages so far. **Recently we officially

Reviewer 4: ```suggestion
 the team added React Testing Library to `package.json`. We have both packages so far. **Recently we officially
```
Usunęłabym przedimek, dodajemy konkretny byt.

>   decided to slowly rewrite the Enzyme library for good.** We added the rule to our Declaration of Code Convention

Reviewer 4: ```suggestion
  decided to slowly rewrite the Enzyme tests and get rid of the library for good.** We added the rule to our Declaration of Code Convention
```

Może coś takiego? 
Nie przepisujemy samego Enzyme'a, ale napisane w nim testy, chociaż to Enzyme'a chcemy się pozbyć 😉

   that the new components are tested just in RTL, the old ones will be rewritten during the development of
    specific parts.

## Migration effects so far

Currently we have got **592 test files** (counting after the file names with “spec” phrase). In these files we can also

>  find **709 of unique “describe”** usage for wrapping a series of tests, and **1796 unique test cases**. Half of tests

Reviewer: find 709 unique occurences of describe used for wrapping?

>   are written only in Jest, but a significant amount of tests for better testing use extra libraries.

Reviewer 3: ```suggestion
  are written only in Jest, but a significant number of tests for better testing use extra libraries.
```
amount jest do rzeczowników niepoliczalnych

author: [magdalena.mazur]
tags: [tech, frontend, javascript, jest, enzyme, react-testing-library, rtl, react]
---

Everyone repeats like a mantra that tests are an indispensable element of development work. Is there anything to it?
 Well, I need to admit that as a developer, I rather often want to skip the test writing stage. I assume I’m not the
  only one. I’m aware that it’s a mistake, even with testers on board. Effective and efficient testing of your own
   code can help with catching bugs in new functionalities, as well as in changes to already existing ones. It cannot
    be questioned. Sometimes tests also help to understand how some long-unused functionality or component works. And
     that’s a small bonus too. **Can a large project cope without testing? Probably so. But the number of errors and

>       their severity will probably be much higher.** That’s why in Allegro Ads we pay attention to writing tests.

Reviewer 2: to się porobiły jakieś dziwne wcięcia (i w akapitach niżej podobnie)

![Enzyme and RTL dowloands graph](/img/articles/2021-12-13-why-we-should-rewrite-enzymejs-to-rtl-and-how-to-do-that/06.png)
*Source [npm trends](https://www.npmtrends.com/@testing-library/react-vs-enzyme)*

In the case of RTL, it is clear that the number of downloads is still growing strongly. On the other hand, it can be
 suspected that the highest peak of Enzyme downloads is already behind.

## Reasons to change the testing library

Well, let’s be honest, Enzyme.js is a powerful library which has helped us test our React Components in an easier way
 for a long time. But maybe... too long? The first official release was in December 2015. Do not get me wrong, I don’t
  want to say that everything old is bad. But the frontend world is changing constantly, just as the approach to
   testing it. I know that the Enzyme.js library hasn’t been deprecated yet, but there are a lot of reasons to switch
    your test library right now. **If you haven’t read it I highly recommend you to take a look at
     [this article](https://www.piotrstaniow.pl/goodbye-enzyme) by Piotr Staniów** (it takes about 10 minutes to read).

> The main reasons for me to think about slowly rewriting my tests and switch to another library from this article are:

Reviewer 2: switch -> switching
